pkg/stacks/unpack.go

```diff
@@ -17,320 +17,439 @@ limitations under the License.
 package stacks
 
 import (
-	"bytes"
-	"encoding/base64"
 	"encoding/json"
 	"fmt"
 	"io"
-	"io/ioutil"
-	"mime"
-	"os"
 	"path/filepath"
+	"sort"
 	"strings"
 
 	"github.com/ghodss/yaml"
-	"github.com/spf13/afero"
 	appsv1 "k8s.io/api/apps/v1"
 	batchv1 "k8s.io/api/batch/v1"
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 
 	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
 	"github.com/crossplaneio/crossplane/apis/stacks/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/stacks/walker"
 )
 
 const (
-	registryDirName         = ".registry"
-	resourcesDirName        = "resources"
 	installFileName         = "install.yaml"
+	resourceFileNamePattern = "*resource.yaml"
+	groupFileName           = "group.yaml"
 	appFileName             = "app.yaml"
 	permissionsFileName     = "rbac.yaml"
 	iconFileNamePattern     = "icon.*"
-	resourceFileNamePattern = "*crd.yaml"
+	crdFileNamePattern      = "*crd.yaml"
+	uiSpecFileNamePattern   = "*ui-schema.yaml"
+	uiSpecFileName          = "ui-schema.yaml" // global ui schema filename
 	yamlSeparator           = "\n---\n"
+
+	// Stack annotation constants
+	annotationStackIcon           = "stacks.crossplane.io/icon"
+	annotationStackUISpec         = "stacks.crossplane.io/ui-spec"
+	annotationStackTitle          = "stacks.crossplane.io/stack-title"
+	annotationGroupTitle          = "stacks.crossplane.io/group-title"
+	annotationGroupCategory       = "stacks.crossplane.io/group-category"
+	annotationGroupDescription    = "stacks.crossplane.io/group-description"
+	annotationResourceTitle       = "stacks.crossplane.io/resource-title"
+	annotationResourceTitlePlural = "stacks.crossplane.io/resource-title-plural"
+	annotationResourceCategory    = "stacks.crossplane.io/resource-category"
+	annotationResourceDescription = "stacks.crossplane.io/resource-description"
+	annotationKubernetesManagedBy = "app.kubernetes.io/managed-by"
 )
 
 var (
 	log = logging.Logger.WithName("stacks")
 )
 
-// Unpack unpacks the stack contents from the given directory.
-func Unpack(contentDir string) error {
-	log.V(logging.Debug).Info("Unpacking stack", "contentDir", contentDir)
-	fs := afero.NewOsFs()
+// StackResource provides the Stack meta-data for a CRD. This is the format for resource.yaml files.
```

Comment:
nit: can we be consistent with our usage of the word metadata? i don't think i've ever seen it spelled `meta-data` before, and if we are to later on search the code base for `metadata` (for some reason), these instances would not be found in the results.

there are a few instances of `meta-data` in this PR.

Reply:
I was never confident in `metadata` being a real word.  180m to 6.5m Google results. metadata wins.

> The short answer is there is no difference – it is simply a matter of the correct spelling of “metadata“. Call it what you will, taxonomy, meta data or metainformation, it is all metadata.

Nevermind the URL I snagged that SEO drivel from :newspaper:

Reply:
[Resolved](https://github.com/crossplaneio/crossplane/compare/9f5ab3c6d421d5e1b0d722af3317569614d5132f..017c3d47090daca29f82dfd30b95578a3b3e297d)